### ISSUE TITLE
Allow to specify filename (glob) per item in no-restricted-syntax

### ISSUE THREAD

**Issue by JuroOravec** [eslint/eslint#13836]
**What rule do you want to change?**

`no-restricted-syntax`

**Does this change cause the rule to produce more or fewer warnings?**

Neither, but it will provide more flexibility.

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option per syntax item to enable specifying which files the item should apply.

The implementation would affect this file https://github.com/eslint/eslint/blob/3734a669983de7d5107ba8f39b291c6e3116489f/lib/rules/no-restricted-syntax.js

<details>
<summary>
Example Implementation
</summary>

```js
create(context) {
  let filename; // THIS LINE IS NEW
  return context.options.reduce((result, selectorOrObject) => {
      const isStringFormat = (typeof selectorOrObject === "string");
      const hasCustomMessage = !isStringFormat && Boolean(selectorOrObject.message);

      const selector = isStringFormat ? selectorOrObject : selectorOrObject.selector;
      const message = hasCustomMessage ? selectorOrObject.message : `Using '${selector}' is not allowed.`;
      const filenameGlob = !isStringFormat && selectorOrObject.filename;

      // NEW BLOCK START
      // Ignore the selector if filename is specified but it doesn't match the current one
      if (typeof filenameGlob === 'string') {
        // Lazy-load filename
        if (!filename) {
          filename = context.getFilename();
        }
        if (!minimatch(filename, filenameGlob)) return result; // NOTE: ESLint already uses minimatch
      }
      // NEW BLOCK END

      return Object.assign(result, {
          [selector](node) {
              context.report({
                  node,
                  messageId: "restrictedSyntax",
                  data: { message }
              });
          }
      });
  }, {});

}
```

</details>

**Please provide some example code that this change will affect:**

Instead of having all syntax items applied for all files:

<details>

<summary>Current no-restricted-syntax</summary>

```js
    'no-restricted-syntax': [
      'error',
      {
        // Raise on `Vue.extend()`
        selector:
          'CallExpression > MemberExpression[object.name="Vue"][property.name="extend"]',
        message: "Use 'defineComponent' instead of 'Vue.extend'.",
      },
      {
        // Raise when calling watch() inside a watch() callback
        selector:
          'CallExpression[callee.name="watch"] CallExpression[callee.name="watch"]',
        message: 'Do not use nested watchers.',
      },
    ],
```

</details>

The suggestion is to allow to specify filename *per syntax item*. 

The `filename` property would be optional. If not set (`undefined`), that syntax item would be applied to all files.

The value of `filename` property would be a string that would be resolved as a glob pattern.

In other words, the syntax item would be extended by following:

`{ filename?: string }`

This change would be backward-compatible:
- If filename not set, everything is as before
- If the syntax item is set only as a string, everything is as before.

<details>

<summary>Suggested change to no-restricted-syntax</summary>

```js
    'no-restricted-syntax': [
      'error',
      // This rule does not specify the filename, so it is active for all files
      {
        // Raise on `Vue.extend()`
        selector:
          'CallExpression > MemberExpression[object.name="Vue"][property.name="extend"]',
        message: "Use 'defineComponent' instead of 'Vue.extend'.",
      },
      // This rules is active only for 'src/my/happy/file.js'
      {
        // Raise when calling watch() inside a watch() callback
        selector:
          'CallExpression[callee.name="watch"] CallExpression[callee.name="watch"]',
        message: 'Do not use nested watchers.',
  		filename: 'src/my/happy/file.js', // EXACT FILENAME GIVEN
      },
      // This rule is active for any file `.js` in `src/my/happy/*` 
      {
        selector: "CallExpression[callee.name='setTimeout'][arguments.length!=2]",
        message: "setTimeout must always be invoked with two arguments.",
  		filename: 'src/my/happy/*.js', // GLOB FILENAME GIVEN
      }
    ],
```

</details>

**What does the rule currently do for this code?**

The rule is great way to quickly write custom syntax rules without needing to write a custom rules. However, in projects with multiple file types, where different conventions hold for these different filetypes, being able to filter the activity by filename would be useful.

Example: I want to raise error when a variable with a specific name is exported from a `.vue` file, but I want it to pass if the variable with same name is exported from a `.ts` file. The reason being that that variable should be defined outside of `.vue` to allow for proper type inference. 

**What will the rule do after it's changed?**

The rule will allow to specify which filenames the restricted syntax should apply to, per syntax item.

**Are you willing to submit a pull request to implement this change?**

Yes

**Comment by mdjermanovic** [comment]
@JuroOravec thanks for the issue!

> However, in projects with multiple file types, where different conventions hold for these different filetypes, being able to filter the activity by filename would be useful.

Different configurations for different files are already supported in config files: [Configuration Based on Glob Patterns](https://eslint.org/docs/user-guide/configuring#configuration-based-on-glob-patterns).

The following `.eslintrc.js` configuration should match the functionality from the suggested change example:

```js
const common = [
  {
    selector:
      'CallExpression > MemberExpression[object.name="Vue"][property.name="extend"]',
    message: "Use 'defineComponent' instead of 'Vue.extend'.",
  },
];

const srcMyHappy = [
  ...common,
  {
    selector: "CallExpression[callee.name='setTimeout'][arguments.length!=2]",
    message: "setTimeout must always be invoked with two arguments.",
  },
];

const srcMyHappyFile = [
  ...srcMyHappy,
  {
    selector:
      'CallExpression[callee.name="watch"] CallExpression[callee.name="watch"]',
    message: "Do not use nested watchers.",
  },
];

module.exports = {
  rules: {
    "no-restricted-syntax": ["error", ...common],
  },
  overrides: [
    {
      files: ["src/my/happy/*.js"],
      rules: {
        "no-restricted-syntax": ["error", ...srcMyHappy],
      },
    },
    {
      files: ["src/my/happy/file.js"],
      rules: {
        "no-restricted-syntax": ["error", ...srcMyHappyFile],
      },
    },
  ],
};
```

Since this feature already exists at the config level, I think that adding it to individual rules would be redundant, and might be confusing.

**Comment by JuroOravec** [comment]
Hi @mdjermanovic, thanks for reply.

Agree, makes complete sense from the perspective of ESLint. I'll deal with my use case either using the overrides or as a custom rule, thanks for a suggestion.

Closing this issue.
